extending-the-rest-api/controller-classes.md

```diff
@@ -1,14 +1,50 @@
 
+You can avoid this issue by naming your callback functions using a unique prefix such as `myplugin_myendpoint_` to avoid any potential conflics:
+
+```php
+function myplugin_myendpoint_register_routes() { /* ... */ }
+function myplugin_myendpoint_get_item() { /* ... */ }
+function myplugin_myendpoint_get_item_schema() { /* ... */ }
+function myplugin_myendpoint_get_item_permissions_check() { /* ... */ }
+function myplugin_myendpoint_get_items() { /* ... */ }
+function myplugin_myendpoint_get_items_permissions_check() { /* ... */ }
+function myplugin_myendpoint_prepare_item_for_response() { /* ... */ }
+
+add_action( 'rest_api_init', 'myplugin_register_routes' );
+```
+
+You may already be familiar with this approach because it is commonly used within theme `functions.php` files. However these prefixes are unnecessarily verbose, and several better options exist to group and encapsulate your endpoint's logic in a more maintainable way.
+
+### Namespaces _vs_ Classes
+
+WordPress currently requires PHP 5.6 or greater. PHP 5.6 supports <a href="https://www.php.net/manual/en/language.namespaces.rationale.php">namespaces</a>, which provide an easy way to encapsulate your endpoint's functionality. By declaring a `namespace` at the top of your endpoint's PHP file, all methods within that namespace will be declared within that namespace and will no longer conflict with global functions. You may then use shorter, more-readable names for your endpoint callbacks:
+
+```php
+namespace MyPlugin\API\MyEndpoint;
+
+function register_routes() { /* ... */ }
+function get_item() { /* ... */ }
+function get_item_schema() { /* ... */ }
+function get_item_permissions_check() { /* ... */ }
+function get_items() { /* ... */ }
+function get_items_permissions_check() { /* ... */ }
+function prepare_item_for_response() { /* ... */ }
+
+add_action( 'rest_api_init', __NAMESPACE__ . '\\register_routes' );
+```
+
+Before WordPress required a version of PHP which supported namespaces, it was common to declare these related methods in a controller class. Namespaces and controller classes are both robust, maintainable ways to group your endpoint code, and neither is recommended above the other. You may use the approach that best matches the rest of your plugin's codebase.
+
+The remainder of this page details how to write your own controller class.
 
 ## Controllers
 
-Controllers typically do one thing; they receive input, and generate output. For the WordPress REST API our controllers will handle request input as `WP_REST_Request` objects and generate response output as `WP_REST_Response` objects. Let's look at an example controller class:
+Controllers typically do one thing: receive input, and generate output. For the WordPress REST API our controllers will handle request input as `WP_REST_Request` objects and generate response output as `WP_REST_Response` objects. Let's look at an example controller class:
 
 ```php
 class My_REST_Posts_Controller {
```

Comment:
I think my inclination to explain namespaces is that, since in this example we do _not_ extend the base class, we don't get _any_ of the actual benefits of a class— the class is purely used for consolidation of code. In that case I'd generally use a namespace instead, because to my mind a namespace groups code and a class should only be used if it provides other benefits.

Now, a class _does_ provide benefits; we can more easily cache schema (which you added to this example already), we can extend the base controller to get more functionality, etc. But that was my thinking for trying to present namespaces as an alternative. Thanks for your perspective, I'll clean that up a bit.